scheduler/src/cook/mesos/util.clj

```diff
@@ -251,9 +249,42 @@
          (map (partial d/entity db))
          (filter #(<= (.getTime start) (.getTime (:job/submit-time %))))
          (filter #(< (.getTime (:job/submit-time %)) (.getTime end)))
-         (filter #(contains? states (:job/state %)))
+         (filter #(= :job.state/completed (:job/state %)))
+         (filter #(not (:job/custom-executor %)))
+         (take limit))))
+
+(defn get-active-jobs-by-user-and-state
+  "Returns all jobs for a particular user in the specified state
+   and timeframe, without a custom executor.
+   Note that this query is not performant for completed jobs, use
+   get-completed-jobs-by-user instead."
+  [db user start end state]
+  (->> (q '[:find [?j ...]
+            :in $ ?user ?state ?start ?end
+            :where
+            [?j :job/state ?state]
+            [?j :job/user ?user]
+            [?j :job/submit-time ?t]
+            [(<= ?start ?t)]
+            [(< ?t ?end)]
+            [?j :job/custom-executor false]]
+          db user state start end)
+       (map (partial d/entity db))))
+
+(defn get-jobs-by-user-and-states
+  "Returns all jobs for a particular user in the specified states
+   and timeframe, without a custom executor."
+  [db user states start end limit]
+  (let [get-jobs-by-state (fn get-jobs-by-state [state]
+                            (if (= state :job.state/completed)
+                              (get-completed-jobs-by-user db user start end limit)
+                              (get-active-jobs-by-user-and-state db user start end state)))
```

Comment:
The issue is that the time based scanning is generally worse for running jobs because they'll be at the end of the time range. You can do a different query with a similar principal to look for the first job in a given state, but that performed approximately the same as the simple datomic query for my tests (a few users with a ton of jobs and a few users with few jobs) so I don't think it's worth having the more complicated code.